### PR TITLE
improve performance

### PR DESCRIPTION
**Pull request by Uzlopak**
Will provide benchmarks later if requested, but it is about 10 % faster (before 2600 k requests/s, after 2800k requests/s)
It would be nice, if you give me feedback regarding the code.

### CODE REVIEW

**Comment by Uzlopak** [comment]
@mcollina 
Please take a look

File: vary.js
Comment (Uzlopak) [inline vary.js:7]: I wonder if this is actually error prone and maybe should be removed if header can only be a string. We need to actually lowercase the values of the the header. Maybe converting the array with a join to a string and reparse it makes more sense.

**Comment by Uzlopak** [comment]
@mcollina 
@jsumners 
@Eomm 

Does it make sense to create a new package @fastify/vary to use this code also in @fastify/sensible ?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
why do we need vary in sensible?

**Comment by Uzlopak** [comment]
@mcollina 

Sensible also uses vary.

https://github.com/fastify/fastify-sensible/blob/92139a8433b376294f6a3eee51ab38cdb3b3c872/package.json#L39

File: vary.js
Comment (mcollina) [inline vary.js:31]: How much of this code is extracted from the vary module?

**Comment by mcollina** [comment]
It's just exposed to the end users. The API of what was added in here it's not the same of `vary` module so it's not a 100% swap.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: vary.js
Comment (Uzlopak) [reply vary.js:31]: fieldNameRE is from the vary module and the TypeError is transplanted from vary to ensure that it is backwards compatible. Other than that nothing.

Comment (mcollina) [reply vary.js:31]: Please add the license disclaimer of that module in this file

**Comment by Uzlopak** [comment]
I added benchmarks:

```
vary - field to * x 11,460,799 ops/sec ±0.65% (190 runs sampled)
vary - * to field x 8,176,435 ops/sec ±0.92% (187 runs sampled)
vary - field to empty x 6,908,442 ops/sec ±0.77% (188 runs sampled)
vary - fields string to empty x 8,653,009 ops/sec ±0.60% (189 runs sampled)
vary - field to fields x 3,660,382 ops/sec ±0.69% (189 runs sampled)

cors - field to * x 76,832,035 ops/sec ±1.12% (187 runs sampled)
cors - * to field x 34,669,727 ops/sec ±0.95% (186 runs sampled)
cors - field to empty x 27,483,166 ops/sec ±0.51% (191 runs sampled)
cors - fields string to empty x 33,894,136 ops/sec ±1.82% (179 runs sampled)
cors - field to fields x 334,247 ops/sec ±0.19% (194 runs sampled)
```

File: vary.js
Comment (Uzlopak) [reply vary.js:38]: The code snippet I copied is basically 1:1 copied from the RFC. So I referenced the RFC. I changed Error-Messages.

So basically there is no copyrighted material in my PR, as it is based on the RFC.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Uzlopak** [comment]
While running the benchmarks, I realize that the case `cors - field to fields` is 10 times sllower. I will investigate this further.

Sry :(

**Review by climba03003 — CHANGES_REQUESTED** [review]
> While running the benchmarks, I realize that the case cors - field to fields is 10 times sllower. I will investigate this further.

The Request Change here is blocking to land based on the message above.
Let me know when the slow case fix and I will dismiss the Request Change.

File: package.json
Comment (jsumners) [inline package.json:48]: This seems like a really heavy dependency. It also isn't at a stable release number.

**Comment by Uzlopak** [comment]
@jsumners 

tiny-lru is slower than mnemonist

vary:
```
vary - field to undefined x 7,098,732 ops/sec ±0.75% (190 runs sampled)
vary - field to * x 11,084,521 ops/sec ±0.93% (191 runs sampled)
vary - * to field x 7,882,434 ops/sec ±0.69% (188 runs sampled)
vary - field to empty x 6,949,372 ops/sec ±0.77% (189 runs sampled)
vary - fields string to empty x 8,385,889 ops/sec ±0.70% (189 runs sampled)
vary - field to fields x 3,548,317 ops/sec ±0.71% (191 runs sampled)
```

mnemonist:
```
cors - field to undefined x 108,076,176 ops/sec ±2.34% (167 runs sampled)
cors - field to * x 87,009,887 ops/sec ±1.46% (175 runs sampled)
cors - * to field x 79,040,526 ops/sec ±0.40% (193 runs sampled)
cors - field to empty x 76,961,478 ops/sec ±0.53% (191 runs sampled)
cors - fields string to empty x 77,507,047 ops/sec ±0.46% (191 runs sampled)
cors - field to fields x 62,389,119 ops/sec ±1.21% (176 runs sampled)
```

tiny-lru:
```
cors - field to undefined x 106,751,509 ops/sec ±1.82% (171 runs sampled)
cors - field to * x 88,892,481 ops/sec ±1.48% (178 runs sampled)
cors - * to field x 78,700,246 ops/sec ±0.46% (191 runs sampled)
cors - field to empty x 77,239,556 ops/sec ±0.52% (189 runs sampled)
cors - fields string to empty x 76,970,667 ops/sec ±0.61% (191 runs sampled)
cors - field to fields x 27,534,170 ops/sec ±0.55% (191 runs sampled)
```

It is still about 10 times faster than original vary. So if you want I can use tiny-lru

**Comment by mcollina** [comment]
mnemonist is faster than tiny-lru in all possible cases: https://github.com/dominictarr/bench-lru.

**Review by Eomm — APPROVED** [review]

**Review by Fdawgs — APPROVED** [review]

**Comment by kibertoad** [comment]
I suggest pinning mnemonist at patch version, since it's not at 1.0.0 yet, so I expect breaking changes in minor versions.

**Comment by Uzlopak** [comment]
@kibertoad 
pinned

**Comment by Uzlopak** [comment]
@climba03003 

You want to review?

**Review by climba03003 — APPROVED** [review]

**Review by kibertoad — APPROVED** [review]

**Comment by kurtextrem** [comment]
out of curiosity, did anyone benchmark other LRU implementations? (if so, I don't wanna repeat the work)

Mainly asking, because e.g. https://github.com/lukeed/flru is a muuuch smaller dep. (it has a benchmark at the bottom, so maybe those are also worth benchmarking). Here's even a native addon, which might also improve perf of the LRU: https://github.com/adzerk/node-lru-native

**Comment by Uzlopak** [comment]
@kurtextrem 

1. I am actually thinking alot about making a general decision regarding a LRU by fastify so that we use everyhere the same LRU
2. I dont think that we should upstream anything to vary. The caching is something I think would not pass by their maintainers. And tbh I dont expect caching in modules from jshttp-org.
3. I dont think that it has any increadible memory increase. The Vary header is something very deterministic set by the server. So I strongly assume that in production we actually cache about 20 entries in the lru-cache. Its not like we produces masses of different vary headers on the fly.

**Comment by kurtextrem** [comment]
@Uzlopak 
Regarding point 2: what I meant is, this PR changes the algorithm of parsing the `Vary` header in general (excluding the LRU Cache here). The two implementations totally differ now, and I wonder which implementation is faster, since the new implementation benchmarks "new algorithm + LRU" and not "only new algorithm". So what if "old algorithm" (from the `vary` npm module) is faster with the LRU, and if it isn't, then pushing the new algorithm to the package might make sense.

Regarding point 3: In this case, I think we could do better than reserving 1000 slots by default (https://github.com/fastify/fastify-cors/pull/214/files#diff-6e4256d6cce0fe10bc30208d12a22415f2db8cefe0aeebf6fa2830e590539894R70)

**Comment by Uzlopak** [comment]
The only thing we used vary was to add the necessary key to the vary field was not already set. Also of course the special case of *. 

I dont even think that it would actually hurt to have duplicate keys in vary. 

So the only chokepoint is to determine if the key is already in the provided string. I think without caching we have already some noticable gains. I think maybe i can squeeze more perf...

We could reduce cacheSize to 100 if you like. 100 like in 
https://github.com/fastify/fastify/blob/37ae4ea1f7490e8194e42b1bf85a9db60f048c57/lib/contentTypeParser.js#L40

I am actually kind of unopinionated regarding lru caching. I am

**Pull request merged**
